test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1

```diff
@@ -256,6 +260,17 @@ Describe "TabCompletion" -Tags CI {
             Pop-Location
         }
 
+        It "TabCompletion should be case-insensitive for file names" {
```

Comment:
I'd also love to see tests for the order of the completed options. The first one should be the one with a correct casing, if accessible. Otherwise, the UX would be pretty bad in the multiple casing cases.

Reply:
We use alphabetical [sort](https://github.com/iSazonov/PowerShell/blob/d5673cb661bed2d6406cd4e07757db7b927d5528/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L6735) with ignoring casing. I don't know how put an object with correct casing on first place. It looks you request creating very complex sorting.
Do anybody _really_ create multiple files/directories with different casing on Unix?

Reply:
It is very common to have both `Makefile` and `makefile`.

Reply:
It is only one extra <tab> press.
I have no idea how to make this weighted sort by simple way 😕

Reply:
Not exactly simple, but sort insensitive, then compare each element to it's successor in a case sensitive manner, swapping to get case correct first.

Reply:
Yes, my concern is that we will significantly increase the number of comparisons - is it ok for interactive scenarios?

Reply:
> It is only one extra press.

It's not hard to save this extra press for many many people over years to come, is it?

> Yes, my concern is that we will significantly increase the number of comparisons

PowerShell does sooo much stuff behind the scene, it's peanuts compare to it.

Reply:
Sample 1:
wordToComplete = "aB"

Input:
abcd
aBcd
abde
aBde

Result:
aBcd
aBde
abcd
abde

Sample 2:
wordToComplete = "aB*D"

Input:
abcdef
aBCDef
abcdgh
aBcDgh

Result:
aBCDef
aBcDgh
abcdef
abcdgh

-- If we want the results the sort is not simple as seems - we must take into account wildcards.

Reply:
Maybe:
1. Get a case-sensitive array and sort it
2. Get a case-insensitive array and sort it
3. Remove elements of (1) array remove from (2) array
4. Add elements of (2) array after end of (1) array

Reply:
I would:

1. Case-insensitive array sort
2. For each element, compare case **insensitive** with the next element, if they are equal (should be rare), do a case **sensitive** wildcard match with the **original** word to complete to determine the ordering of these 2 elements, swapping if the first fails to match and the second does.

This should be O(n) with an upper bound of 2n, so it's cheaper than 2 sorts.

Reply:
Is my sample 2 right? If so your suggestion don't work for it. We get:
```
Result:
aBcd
abcd
aBde
abde

```

Reply:
I'm not sure there is a right and wrong result here - only a least surprising, and I think my suggestion prefers case insensitive over the wildcard match which is less surprising, but still orders with some case sensitivity.

At any rate, real world experience will guide us best - try and keep it simple and handle the `Makefile`/`makefile` case nicely, and we'll see if it's sufficient.